### PR TITLE
OpenSSL 3.0: ECDSA signer

### PR DESCRIPTION
**Pull request by fredmorcos**
### Short description
Support OpenSSL 3.0 in the ECDSA signer.

### Checklist
I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [X] added or modified unit test(s)

### CODE REVIEW

File: pdns/opensslsigners.cc [hunk 1]
```diff
@@ -1285,24 +1571,61 @@ void OpenSSLECDSADNSCryptoKeyEngine::fromISCMap(DNSKEYRecordContent& drc, std::m
...
-  int ret = EC_KEY_set_private_key(d_eckey.get(), prv_key.get());
+  std::string publicKey{};
+  size_t bufsize = EC_POINT_point2oct(d_group.get(), publicKeyECPoint.get(), POINT_CONVERSION_COMPRESSED, nullptr, 0, nullptr);
+  if (bufsize == 0) {
+    throw pdns::OpenSSL::error(getName(), "Failed to get public key binary buffer size");
+  }
+  publicKey.resize(bufsize);
+
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
+  auto* publicKeyCStr = const_cast<unsigned char*>(reinterpret_cast<const unsigned char*>(publicKey.c_str()));
```
Comment (fredmorcos) [inline pdns/opensslsigners.cc:1594]: Argh.. I missed this one `.data()` instead of `.c_str()`.

File: pdns/opensslsigners.cc [hunk 2]
```diff
@@ -1233,39 +1408,118 @@ std::string OpenSSLECDSADNSCryptoKeyEngine::sign(const std::string& message) con
...
 {
-  if (signature.length() != (d_len * 2)) {
+  if (signature.length() != (static_cast<unsigned long>(d_len) * 2)) {
     throw runtime_error(getName() + " invalid signature size " + std::to_string(signature.length()));
   }
 
-  string l_hash = this->hash(message);
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
+  auto* signatureCStr = const_cast<unsigned char*>(reinterpret_cast<const unsigned char*>(signature.c_str()));
+  auto rComponent = BigNum(BN_bin2bn(signatureCStr, d_len, nullptr), BN_clear_free);
```
Comment (rgacogne) [inline pdns/opensslsigners.cc:1417]: Given that this is a signature (not secret) we could use `BN_free` instead to avoid zeroing the memory.

**Review by rgacogne — APPROVED** [review]
Great! A few nits but nothing that would prevent this PR from being merged as far as I4m concerned.

File: pdns/opensslsigners.cc [hunk 3]
```diff
@@ -19,6 +19,13 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
  */
+#include "base64.hh"
```
[the remark this answers is not in the document]
Comment (fredmorcos) [reply pdns/opensslsigners.cc:22]: No, a debugging relic.

File: pdns/opensslsigners.cc [hunk 4]
```diff
@@ -19,6 +19,13 @@
...
+#include "misc.hh"
+#include <cstdio>
```
[the remark this answers is not in the document]
Comment (fredmorcos) [reply pdns/opensslsigners.cc:24]: Also a debugging relic.

File: pdns/opensslsigners.cc [hunk 5]
```diff
@@ -1033,95 +1040,154 @@ class OpenSSLECDSADNSCryptoKeyEngine : public DNSCryptoKeyEngine
...
+  using KeyContext = std::unique_ptr<EVP_PKEY_CTX, decltype(&EVP_PKEY_CTX_free)>;
+  using MessageDigestContext = std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_free)>;
   using Group = std::unique_ptr<EC_GROUP, decltype(&EC_GROUP_free)>;
   using Point = std::unique_ptr<EC_POINT, decltype(&EC_POINT_free)>;
   using Signature = std::unique_ptr<ECDSA_SIG, decltype(&ECDSA_SIG_free)>;
 
-  unsigned int d_len;
+  int d_id{0};
+  int d_len{0};
+  const char* d_group_name{nullptr};
```
[the remark this answers is not in the document]
Comment (fredmorcos) [reply pdns/opensslsigners.cc:1076]: Right. We had discussed this for the RSA signer and I forgot to update it for ECDSA.

**Comment by fredmorcos** [comment]
Let me know before merging so that I squash the fixups.

File: pdns/opensslsigners.cc [hunk 6]
```diff
@@ -1233,39 +1406,118 @@ std::string OpenSSLECDSADNSCryptoKeyEngine::sign(const std::string& message) con
...
-  if (signature.length() != (d_len * 2)) {
+  if (signature.length() != (static_cast<unsigned long>(d_len) * 2)) {
     throw runtime_error(getName() + " invalid signature size " + std::to_string(signature.length()));
   }
 
-  string l_hash = this->hash(message);
+  // NOLINTNEXTLINE(*-cast): Using OpenSSL C APIs.
+  auto* signatureCStr = const_cast<unsigned char*>(reinterpret_cast<const unsigned char*>(signature.c_str()));
+  auto rComponent = BigNum(BN_bin2bn(signatureCStr, d_len, nullptr), BN_free);
+  auto sComponent = BigNum(BN_bin2bn(signatureCStr + d_len, d_len, nullptr), BN_clear_free);
```
Comment (rgacogne) [inline pdns/opensslsigners.cc:1416]:
```suggestion
  auto sComponent = BigNum(BN_bin2bn(signatureCStr + d_len, d_len, nullptr), BN_free);
```

**Review by rgacogne — APPROVED** [review]
Looks good, thanks! One small remaining nit, I think you are good to squash :)

**Comment by fredmorcos** [comment]
Reminder to merge :-)

**Pull request merged**